GVFS/GVFS.Upgrader/UpgradeOrchestrator.cs

```diff
@@ -0,0 +1,389 @@
+                    Version newVersion = null;
+                    if (!this.TryRunUpgradeInstall(out newVersion, out error))
+                    {
+                        this.ExitCode = ReturnCode.GenericError;
+                    }
+                }
+                finally
+                {
+                    string remountError = null;
+                    if (!this.TryRemountRepositories(out remountError))
+                    {
+                        remountError = Environment.NewLine + "WARNING: " + remountError;
+                        this.output.WriteLine(remountError);
+                        this.ExitCode = ReturnCode.Success;
+                    }
+
+                    this.DeletedDownloadedAssets();
+                }
+            }
+            
+            if (this.ExitCode == ReturnCode.GenericError)
+            {
+                error = Environment.NewLine + "ERROR: " + error;
+                this.output.WriteLine(error);
+            }
+
+            if (this.input == Console.In)
+            {
+                this.output.WriteLine("Press Enter to exit.");
+                this.input.ReadLine();
+            }
+            
+            if (this.shouldExitOnError)
+            {
+                Environment.Exit((int)this.ExitCode);
+            }
+        }
+
+        private bool LaunchInsideSpinner(Func<bool> method, string message)
+        {
+            return ConsoleHelper.ShowStatusWhileRunning(
+                method,
+                message,
+                this.output,
+                this.output == Console.Out && !GVFSPlatform.Instance.IsConsoleOutputRedirectedToFile(),
+                null);
+        }
+
+        private bool TryRunUpgradeInstall(out Version newVersion, out string consoleError)
+        {
+            newVersion = null;
+
+            Version newGVFSVersion = null;
+            GitVersion newGitVersion = null;
+            string errorMessage = null;
+            if (!this.LaunchInsideSpinner(
+                () =>
+                {
+                    if (!this.TryCheckIfUpgradeAvailable(out newGVFSVersion, out errorMessage) ||
+                        !this.TryGetNewGitVersion(out newGitVersion, out errorMessage))
+                    {
+                        return false;
+                    }
+
+                    this.LogInstalledVersionInfo();
+                    this.LogVersionInfo(newGVFSVersion, newGitVersion, "Available Version");
+
+                    this.preRunChecker.CommandToRerun = "gvfs upgrade --confirm";
+                    if (!this.preRunChecker.TryRunPreUpgradeChecks(out errorMessage))
+                    {
+                        return false;
+                    }
+
+                    if (!this.TryDownloadUpgrade(newGVFSVersion, out errorMessage))
+                    {
+                        return false;
+                    }
+
+                    return true;
+                },
+                "Downloading"))

```

Comment:
This is going to say "downloading" even if we never got to downloading because of the pre-req checks.

Does the call to `TryRunPreUpgradeChecks` need to happen so late in the process? Can't it be the first thing you check, before even asking GitHub for new releases? If we can move `TryRunPreUpgradeChecks` to be the first check, and do it before this spinner, that will result in a cleaner experience.